src/execution/sql/sorter.cpp

```diff
@@ -0,0 +1,343 @@
+    const Sorter *const sorter = tl_sorters[sorter_idx];
+    auto part_size = sorter->NumTuples() / (splitters.size() + 1);
+    for (u32 i = 0; i < splitters.size(); i++) {
+      splitters[i][sorter_idx] = sorter->tuples_[(i + 1) * part_size];
+    }
+  }
+
+  timer.ExitStage();
+
+  // -------------------------------------------------------
+  // 4. Compute work packages
+  // -------------------------------------------------------
+
+  timer.EnterStage("Compute Work Packages");
+
+  // Where the merging work units are collected
+  using SeqType = decltype(tuples_);
+  using SeqTypeIter = SeqType::iterator;
+  using MergeWorkType = MergeWork<SeqTypeIter>;
+  std::vector<MergeWorkType> merge_work;
+
+  {
+    // This tracks the current position in the global output (i.e., this
+    // sorter's tuples vector) where the next merge package will begin writing
+    // results into. It begins at the front; as we generate merge packages, we
+    // calculate the next position by computing the sizes of the merge packages.
+    // We've already perfectly sized the output so this memory is allocated and
+    // ready to be written to.
+    auto write_pos = tuples_.begin();
+
+    // This vector tracks, for each sorter, the position of the start of the
+    // next input range. As we move through the splitters, we bump this pointer
+    // so that we don't need to perform two binary searches to find the lower
+    // and upper range around the splitter key.
+    std::vector<SeqTypeIter> next_start(tl_sorters.size());
+
+    for (u32 idx = 0; idx < splitters.size(); idx++) {
+      // Sort the local separators and choose the median
+      ips4o::sort(splitters[idx].begin(), splitters[idx].end(), comp);
+
+      // Find the median-of-medians splitter key
+      const byte *splitter = splitters[idx][tl_sorters.size() / 2];
+
+      // The vector where we collect all input ranges that feed the merge work
+      std::vector<MergeWork<SeqTypeIter>::Range> input_ranges;
+
+      SeqTypeIter::difference_type part_size = 0;
+      for (u32 sorter_idx = 0; sorter_idx < tl_sorters.size(); sorter_idx++) {
+        // Get the [start,end) range in the current sorter such that
+        // start <= splitter < end
+        Sorter *const sorter = tl_sorters[sorter_idx];
+        auto start = (idx == 0 ? sorter->tuples_.begin() : next_start[sorter_idx]);
+        auto end = sorter->tuples_.end();
+        if (idx < splitters.size() - 1) {
+          end = std::upper_bound(start, end, splitter, comp);
+        }
+
+        // If the the range [start, end) is non-empty, push it in as work
+        if (start != end) {
+          input_ranges.emplace_back(start, end);
+        }
+
+        part_size += (end - start);
+        next_start[sorter_idx] = end;
+      }
+
+      // Add work
+      merge_work.emplace_back(std::move(input_ranges), write_pos);
+
+      // Bump new write position
+      write_pos += part_size;
+    }
+  }
```

Comment:
As best I can tell, the scoping is not used, so we should probably remove the blocks.  If that causes problems, then we should document the scoping problem this code block solves.

Reply:
It's there mostly to demarcate the stages and what variables are needed in each stage. We can probably extract each into a separate function.

Reply:
It only stood out to me because it was the only one demarcated like this.  I'm fine with functions, brackets, or just code blocks as long as we're consistent.